tasks/bitmap/src/lib.rs

```diff
@@ -36,11 +36,76 @@ impl Image {
         let mut writer = BufWriter::new(file);
         writeln!(&mut writer, "P6")?;
         writeln!(&mut writer, "{} {} 255", self.width, self.height)?;
-        for color in &self.data {
-            for &channel in &[color.red, color.green, color.blue] {
-                write!(&mut writer, "{}", channel as u8)?;
-            }
-        }
+        writer.write_all(
+            &self
+                .data
+                .iter()
+                .map(|color| vec![color.red, color.green, color.blue])
+                .flatten()
+                .collect::<Vec<u8>>(),
+        )?;
+        Ok(())
+    }
+
+    pub fn read_ppm(&mut self, filename: &str) -> Result<(), Error> {
+        let file = File::open(filename)?;
+        let mut reader = BufReader::new(file);
+
+        let mut buf = String::new();
+        reader.read_line(&mut buf)?;
+
+        // works only with P6 (binary)
+        assert!(buf == "P6\n");
+
+        buf.clear();
+        reader.read_line(&mut buf)?;
+
+        // header parameters assumed to be separated by spaces
```

Comment:
What do you think about implementing the parsing with a parser combinator library like [nom](https://docs.rs/nom/6.1.2/nom/)? I think it might make the error handling more robust, and would be a cool demonstration.

Reply:
Sounds like an interesting challenge, let me try it.

Reply:
I'm struggling with nom, I have expected a steep learning curve, but the error handling which is really weird. I have a working example, but I'm not really satisfied with that.

Reply:
I can review what you have if you'd like.

Reply:
Thanks in advance for your review - pls see the latest commit. I have added nom based parsing, and also tried to implement both P3 (ascii) and P6 (binary) format.

Reply:
Looks good to me. I think we can remove the non-nom based parsing.

Reply:
OK, removed